File: mcp-stdio-server.js
#!/usr/bin/env node

/**
 * MCP stdio Server Demo - File Operations
 * Demonstrates an MCP server using stdio transport for local file operations
 */

import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

class MCPStdioServer {
    constructor() {
        this.capabilities = {
            tools: {},
            resources: {},
            prompts: {},
            logging: {}
        };
        
        this.tools = new Map();
        this.resources = new Map();
        
        this.setupTools();
        this.setupResources();
        this.setupIO();
    }

    setupTools() {
        // File writing tool
        this.tools.set('write_file', {
            name: 'write_file',
            description: 'Write content to a file',
            inputSchema: {
                type: 'object',
                properties: {
                    path: {
                        type: 'string',
                        description: 'Path to the file to write'
                    },
                    content: {
                        type: 'string',
                        description: 'Content to write to the file'
                    }
                },
                required: ['path', 'content']
            }
        });

        // File reading tool
        this.tools.set('read_file', {
            name: 'read_file',
            description: 'Read content from a file',
            inputSchema: {
                type: 'object',
                properties: {
                    path: {
                        type: 'string',
                        description: 'Path to the file to read'
                    }
                },
                required: ['path']
            }
        });

        // List files tool
        this.tools.set('list_files', {
            name: 'list_files',
            description: 'List files in a directory',
            inputSchema: {
                type: 'object',
                properties: {
                    directory: {
                        type: 'string',
                        description: 'Directory path to list files from',
                        default: '.'
                    }
                }
            }
        });

        // Delete file tool
        this.tools.set('delete_file', {
            name: 'delete_file',
            description: 'Delete a file',
            inputSchema: {
                type: 'object',
                properties: {
                    path: {
                        type: 'string',
                        description: 'Path to the file to delete'
                    }
                },
                required: ['path']
            }
        });
    }

    setupResources() {
        // Current directory resource
        this.resources.set('file://current-directory', {
            uri: 'file://current-directory',
            name: 'Current Directory',
            description: 'Files in the current working directory',
            mimeType: 'application/json'
        });
    }

    setupIO() {
        // Handle stdin input
        process.stdin.on('data', (data) => {
            const input = data.toString().trim();
            if (input) {
                try {
                    const request = JSON.parse(input);
                    this.handleRequest(request);
                } catch (error) {
                    this.sendError(null, -32700, 'Parse error', error.message);
                }
            }
        });

        // Handle process termination
        process.on('SIGINT', () => {
            this.log('Server shutting down...');
            process.exit(0);
        });

        process.on('SIGTERM', () => {
            this.log('Server shutting down...');
            process.exit(0);
        });
    }

    async handleRequest(request) {
        const { jsonrpc, id, method, params } = request;

        if (jsonrpc !== '2.0') {
            return this.sendError(id, -32600, 'Invalid Request', 'Invalid JSON-RPC version');
        }

        try {
            switch (method) {
                case 'initialize':
                    await this.handleInitialize(id, params);
                    break;
                case 'notifications/initialized':
                    // No response needed for notifications
                    break;
                case 'tools/list':
                    await this.handleListTools(id);
                    break;
                case 'tools/call':
                    await this.handleCallTool(id, params);
                    break;
                case 'resources/list':
                    await this.handleListResources(id);
                    break;
                case 'resources/read':
                    await this.handleReadResource(id, params);
                    break;
                default:
                    this.sendError(id, -32601, 'Method not found', `Unknown method: ${method}`);
            }
        } catch (error) {
            this.sendError(id, -32603, 'Internal error', error.message);
        }
    }

    async handleInitialize(id, params) {
        this.log('Initializing MCP stdio server...');
        
        const result = {
            protocolVersion: '2024-11-05',
            capabilities: this.capabilities,
            serverInfo: {
                name: 'mcp-stdio-file-server',
                version: '1.0.0'
            }
        };

        this.sendResponse(id, result);
    }

    async handleListTools(id) {
        const tools = Array.from(this.tools.values());
        this.sendResponse(id, { tools });
    }

    async handleCallTool(id, params) {
        const { name, arguments: args } = params;
        
        if (!this.tools.has(name)) {
            return this.sendError(id, -32602, 'Invalid params', `Unknown tool: ${name}`);
        }

        let result;
        
        try {
            switch (name) {
                case 'write_file':
                    result = await this.writeFile(args.path, args.content);
                    break;
                case 'read_file':
                    result = await this.readFile(args.path);
                    break;
                case 'list_files':
                    result = await this.listFiles(args.directory || '.');
                    break;
                case 'delete_file':
                    result = await this.deleteFile(args.path);
                    break;
                default:
                    throw new Error(`Tool ${name} not implemented`);
            }
            
            this.sendResponse(id, {
                content: [
                    {
                        type: 'text',
                        text: JSON.stringify(result, null, 2)
                    }
                ]
            });
        } catch (error) {
            this.sendResponse(id, {
                content: [
                    {
                        type: 'text',
                        text: `Error: ${error.message}`
                    }
                ],
                isError: true
            });
        }
    }

    async handleListResources(id) {
        const resources = Array.from(this.resources.values());
        this.sendResponse(id, { resources });
    }

    async handleReadResource(id, params) {
        const { uri } = params;
        
        if (!this.resources.has(uri)) {
            return this.sendError(id, -32602, 'Invalid params', `Unknown resource: ${uri}`);
        }

        try {
            let content;
            
            if (uri === 'file://current-directory') {
                const files = await this.listFiles('.');
                content = JSON.stringify(files, null, 2);
            }
            
            this.sendResponse(id, {
                contents: [
                    {
                        uri,
                        mimeType: 'application/json',
                        text: content
                    }
                ]
            });
        } catch (error) {
            this.sendError(id, -32603, 'Internal error', error.message);
        }
    }

    // File operation implementations
    async writeFile(filePath, content) {
        const fullPath = path.resolve(filePath);
        await fs.writeFile(fullPath, content, 'utf8');
        return {
            success: true,
            message: `File written successfully: ${filePath}`,
            path: fullPath,
            size: content.length
        };
    }

    async readFile(filePath) {
        const fullPath = path.resolve(filePath);
        const content = await fs.readFile(fullPath, 'utf8');
        const stats = await fs.stat(fullPath);
        
        return {
            success: true,
            path: fullPath,
            content,
            size: stats.size,
            modified: stats.mtime
        };
    }

    async listFiles(directory) {
        const fullPath = path.resolve(directory);
        const entries = await fs.readdir(fullPath, { withFileTypes: true });
        
        const files = [];
        for (const entry of entries) {
            const entryPath = path.join(fullPath, entry.name);
            const stats = await fs.stat(entryPath);
            
            files.push({
                name: entry.name,
                path: entryPath,
                type: entry.isDirectory() ? 'directory' : 'file',
                size: entry.isFile() ? stats.size : null,
                modified: stats.mtime
            });
        }
        
        return {
            success: true,
            directory: fullPath,
            files
        };
    }

    async deleteFile(filePath) {
        const fullPath = path.resolve(filePath);
        await fs.unlink(fullPath);
        
        return {
            success: true,
            message: `File deleted successfully: ${filePath}`,
            path: fullPath
        };
    }

    // Communication helpers
    sendResponse(id, result) {
        const response = {
            jsonrpc: '2.0',
            id,
            result
        };
        
        process.stdout.write(JSON.stringify(response) + '\n');
    }

    sendError(id, code, message, data = null) {
        const response = {
            jsonrpc: '2.0',
            id,
            error: {
                code,
                message,
                ...(data && { data })
            }
        };
        
        process.stdout.write(JSON.stringify(response) + '\n');
    }

    log(message) {
        // Log to stderr so it doesn't interfere with JSON-RPC communication
        process.stderr.write(`[MCP stdio Server] ${message}\n`);
    }
}

// Start the server
const server = new MCPStdioServer();
server.log('MCP stdio File Server started. Waiting for requests...');

File: mcp_stdio_server.py
#!/usr/bin/env python3
"""
MCP Server with stdio transport - File Operations Tool
This server exposes a file operations tool that can read, write, and list files.
"""

from pathlib import Path

from fastmcp import FastMCP

# Create an MCP server
mcp = FastMCP("FileOperationsServer")

@mcp.tool()
def read_file(file_path: str) -> str:
    """
    Read the contents of a file.
    
    Args:
        file_path: Path to the file to read
        
    Returns:
        The contents of the file as a string
    """
    try:
        path = Path(file_path)
        if not path.exists():
            return f"Error: File '{file_path}' does not exist"
        
        if not path.is_file():
            return f"Error: '{file_path}' is not a file"
            
        with open(path, 'r', encoding='utf-8') as f:
            content = f.read()
            
        return f"Contents of '{file_path}':\n{content}"
    except Exception as e:
        return f"Error reading file '{file_path}': {str(e)}"

@mcp.tool()
def write_file(file_path: str, content: str) -> str:
    """
    Write content to a file.
    
    Args:
        file_path: Path to the file to write
        content: Content to write to the file
        
    Returns:
        Success or error message
    """
    try:
        path = Path(file_path)
        
        # Create parent directories if they don't exist
        path.parent.mkdir(parents=True, exist_ok=True)
        
        with open(path, 'w', encoding='utf-8') as f:
            f.write(content)
            
        return f"Successfully wrote {len(content)} characters to '{file_path}'"
    except Exception as e:
        return f"Error writing to file '{file_path}': {str(e)}"

@mcp.tool()
def list_directory(directory_path: str = ".") -> str:
    """
    List the contents of a directory.
    
    Args:
        directory_path: Path to the directory to list (defaults to current directory)
        
    Returns:
        A formatted list of directory contents
    """
    try:
        path = Path(directory_path)
        if not path.exists():
            return f"Error: Directory '{directory_path}' does not exist"
        
        if not path.is_dir():
            return f"Error: '{directory_path}' is not a directory"
            
        items = []
        for item in sorted(path.iterdir()):
            if item.is_dir():
                items.append(f"📁 {item.name}/")
            else:
                size = item.stat().st_size
                items.append(f"📄 {item.name} ({size} bytes)")
                
        if not items:
            return f"Directory '{directory_path}' is empty"
            
        return f"Contents of '{directory_path}':\n" + "\n".join(items)
    except Exception as e:
        return f"Error listing directory '{directory_path}': {str(e)}"

@mcp.resource("file://{file_path}")
def get_file_resource(file_path: str) -> str:
    """
    Get file contents as a resource.
    
    Args:
        file_path: Path to the file
        
    Returns:
        File contents
    """
    try:
        path = Path(file_path)
        if not path.exists() or not path.is_file():
            return f"File '{file_path}' not found or is not a file"
            
        with open(path, 'r', encoding='utf-8') as f:
            return f.read()
    except Exception as e:
        return f"Error reading file: {str(e)}"

@mcp.prompt()
def file_analysis_prompt(file_path: str, analysis_type: str = "summary") -> str:
    """
    Generate a prompt for analyzing a file.
    
    Args:
        file_path: Path to the file to analyze
        analysis_type: Type of analysis (summary, code_review, documentation)
        
    Returns:
        A prompt for file analysis
    """
    analysis_prompts = {
        "summary": f"Please provide a concise summary of the file '{file_path}'. Include the main purpose, key components, and any notable features.",
        "code_review": f"Please perform a code review of the file '{file_path}'. Look for potential issues, improvements, and best practices.",
        "documentation": f"Please generate documentation for the file '{file_path}'. Include function descriptions, usage examples, and any important notes."
    }
    
    return analysis_prompts.get(analysis_type, analysis_prompts["summary"])

if __name__ == "__main__":
    # Run the server with stdio transport (default)
    mcp.run()
